### DISCUSSION TITLE
How to deploy a telegram bot written in Python to Deta.

### DISCUSSION THREAD

**Discussion by anorprogrammer** [deta/cloud-docs#427] [Help]
I am a [telegram bot](https://core.telegram.org/bots/samples#python) developer at python. Telegram bots work on [long polling](https://en.wikipedia.org/wiki/Push_technology#Long_polling) and [webhook](https://en.wikipedia.org/wiki/Webhook). Long polling is used a lot. 

I was using [procfile ](https://devcenter.heroku.com/articles/procfile) to run long polling projects on [heroku](https://heroku.com/)

Procfile:
```procfile
worker: python worker.py
```

Python file: worker.py
```python
import logging

from aiogram import Bot, Dispatcher, executor, types

API_TOKEN = 'BOT TOKEN HERE'

# Configure logging
logging.basicConfig(level=logging.INFO)

# Initialize bot and dispatcher
bot = Bot(token=API_TOKEN)
dp = Dispatcher(bot)


@dp.message_handler(commands=['start', 'help'])
async def send_welcome(message: types.Message):
    """
    This handler will be called when user sends `/start` or `/help` command
    """
    await message.reply("Hi!\nI'm EchoBot!\nPowered by aiogram.")


@dp.message_handler()
async def echo(message: types.Message):
    # echo message
    await message.answer(message.text)


if __name__ == '__main__':
    executor.start_polling(dp, skip_updates=True)
```

How do I start long polling projects in Deta ???

**Comment by scriptyyy13** [comment]
Telegram api works on websockets, which are not supported by deta micro. So, alas, on such an excellent service it’s impossible to host a telegram bot (

**Comment by jnsougata** [comment]
However you can use Telegram Bot's [`webhook`](https://core.telegram.org/bots/api#setwebhook) mode in [Deta Micros](https://docs.deta.sh/docs/micros/deploy#example-code)
That doesn't require websockets and will serve your purpose easily. 
In that way you can receive and parse tg events though wsgi or asgi apps like `FastAPI` `Flask` etc.
![image](https://user-images.githubusercontent.com/53375272/187117522-145a689b-14a8-4bc7-83b8-1f31531a7fa8.png)

**Comment by anorprogrammer** [comment]
@jnsougata 
I used this [example ](https://github.com/aiogram/aiogram/blob/dev-2.x/examples/webhook_example.py) in the webhook but it gave the following error
```console
Traceback (most recent call last):
  File "/opt/python/detalib/debugger.py", line 142, in wrap
    result = func(event, context)
  File "/var/task/_entry.py", line 14, in handler
    return handle(event, main)
  File "/opt/python/detalib/handler.py", line 8, in handle
    assert hasattr(
AssertionError: We couldn't find your 'app'. Please refer to the Manual.
```
Is it possible to upload only projects with app object to **deta** ?

**Reply by jnsougata** [reply]
you must have to name your app variable `app` for example see my bot here https://github.com/jnsougata/yagptb

**Reply by anorprogrammer** [reply]
@jnsougata but your package is based on fastapi.
can you show me the code sample to deploy the [aiohttp ](https://docs.aiohttp.org/en/stable/) project.

**Reply by jnsougata** [reply]
you can not run `aiohttp` asgi app in micros directly for the time beginning… ig

**Comment by nuson999** [comment]
You may want to refer to this code if anyone has the same issue.
I modified the webhook example from [pyTelegramBotAPI](https://github.com/eternnoir/pyTelegramBotAPI)

```
import os

from flask import Flask, request

import telebot

TOKEN = '<your_token>'
bot = telebot.TeleBot(TOKEN)
app = Flask(__name__)


@bot.message_handler(commands=['start'])
def start(message):
    bot.reply_to(message, 'Hello, ' + message.from_user.first_name)


@bot.message_handler(func=lambda message: True, content_types=['text'])
def echo_message(message):
    bot.reply_to(message, message.text)


@app.route('/' + TOKEN, methods=['POST'])
def getMessage():
    json_string = request.get_data().decode('utf-8')
    update = telebot.types.Update.de_json(json_string)
    bot.process_new_updates([update])
    return "!", 200


@app.route("/")
def webhook():
    bot.remove_webhook()
    bot.set_webhook(url='https://example.deta.dev/' + TOKEN)
    return "!", 200


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=int(os.environ.get('PORT', 5000)))
```

**Reply by iamgiolaga** [reply]
As far as you know, is the same approach valid for [python-telegram-bot](https://github.com/python-telegram-bot/python-telegram-bot)?

In my case I have this
```python
if __name__ == '__main__':
    connection = connect()
    updater = Updater(token=TOKEN, use_context=True)
    dispatcher = updater.dispatcher

    echo_handler = MessageHandler(Filters.text & (~Filters.command), echo) # echo is a method that handles Bot answers
    dispatcher.add_handler(echo_handler)

    #updater.start_polling() 
    updater.start_webhook(listen="0.0.0.0", webhook_url=<someUrl + token>, url_path=TOKEN, port=int(os.environ.get('PORT', 5000)))
    updater.idle()
```

As you can see I can run the bot in two "modes": polling and webhook, I would like to keep this opportunity in Deta, as well
